# [E-2] Project 1-2-3

####  순서 : 필요한 모듈 호출 $\rightarrow$ 데이터 준비 $\rightarrow$ 데이터 분리 $\rightarrow$ 모델학습 및 예측 $\rightarrow$ 모델별 비교 평가

## <span style="color:blue">프로젝트-1  load digits : 손글씨 분류</span>

### 1. 필요한 모듈 호출

* 지난 [E1]에 필요한 모듈이 있을 때마다 직전에 호출하는 순서를 따를 경우 넘파이를 이용해 데이터셋을 준비하는 과정에서 이와 관련한 오류 발생
* 이는 본인이 "import"라는 명령어에 대한 이해가 부족한 상태로 프로젝트1의 코드를 짰기 때문으로 판단됨
* 앞으로 수행할 프로젝트의 깔끔하고 명료한 진행을 위하여 이번처럼 전 과정에 걸쳐 공통적으로 필요한 모든 모듈들을 (추후 혼선 방지 차원에서) 우선적으로 불러오는 것에서부터 시작을 하는 것이 좋다고 판단됨  
* 이는 프로젝트 계획 시, 뼈대를 세우는 것과 같다고 여겨짐

In [1]:
# 데이터 준비, 데이터 분리 관련 모듈들
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split

# 데이터 학습 모델들(5가지) : 의사결정트리, 랜덤포레스트, SVM, SGD, 로지스틱회귀
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

# 모델 평가를 위한 오차행렬과 각 지표를 한눈에 확인하기 위한 모듈
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score  # 노드에는 이에 대한 언급 없이 그냥 사용해서, 코드를 그대로 따라하면 오류 발생

### 2. 데이터 준비

#### (1) 데이터셋 가져오기

In [2]:
digits = load_digits()

print(type(dir(digits)))  # dir() = 객체의 변수와 메서드 나열
digits.keys()  # keys() 메서드 = digits 에 담긴 정보 확인

<class 'list'>


dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])

#### (2) 데이터를 손글씨 변수에 저장하고 크기 확인하기

In [3]:
# shape는 배열의 형상정보를 출력
#이 경우, 1797개의 데이터가 64개의 숫자 정보를 담고 있음

digits_data = digits.data
print(digits_data.shape)


# 아래의 코드는 선택사항으로 여겨진다. (저장한 데이터 중 하나 확인해보기 : 어떤 특징으로 분류된 건지 알 수 있음)
digits_data[0]

(1797, 64)


array([ 0.,  0.,  5., 13.,  9.,  1.,  0.,  0.,  0.,  0., 13., 15., 10.,
       15.,  5.,  0.,  0.,  3., 15.,  2.,  0., 11.,  8.,  0.,  0.,  4.,
       12.,  0.,  0.,  8.,  8.,  0.,  0.,  5.,  8.,  0.,  0.,  9.,  8.,
        0.,  0.,  4., 11.,  0.,  1., 12.,  7.,  0.,  0.,  2., 14.,  5.,
       10., 12.,  0.,  0.,  0.,  0.,  6., 13., 10.,  0.,  0.,  0.])

#### (3) 라벨 변수에 타겟 저장

In [4]:
digits_label = digits.target  # 변수에 타겟 저장
print(digits_label.shape)  # 타겟의 길이와 형태 확인 : 총 1797개의 데이터, 라벨값 끝부분이 이상하게 출력됨??
digits_label

(1797,)


array([0, 1, 2, ..., 8, 9, 8])

#### (4) 타겟 이름 출력

In [5]:
digits.target_names  # 10가지로 분류되니까 0~9까지 순서대로 라벨 출력

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

#### (5) 데이터셋 설명 확인

In [6]:
print(digits.DESCR)  # 말그대로 describe

.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 5620
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels are counted in each blo

#### (6) feature 이름 확인

In [7]:
digits.feature_names

['pixel_0_0',
 'pixel_0_1',
 'pixel_0_2',
 'pixel_0_3',
 'pixel_0_4',
 'pixel_0_5',
 'pixel_0_6',
 'pixel_0_7',
 'pixel_1_0',
 'pixel_1_1',
 'pixel_1_2',
 'pixel_1_3',
 'pixel_1_4',
 'pixel_1_5',
 'pixel_1_6',
 'pixel_1_7',
 'pixel_2_0',
 'pixel_2_1',
 'pixel_2_2',
 'pixel_2_3',
 'pixel_2_4',
 'pixel_2_5',
 'pixel_2_6',
 'pixel_2_7',
 'pixel_3_0',
 'pixel_3_1',
 'pixel_3_2',
 'pixel_3_3',
 'pixel_3_4',
 'pixel_3_5',
 'pixel_3_6',
 'pixel_3_7',
 'pixel_4_0',
 'pixel_4_1',
 'pixel_4_2',
 'pixel_4_3',
 'pixel_4_4',
 'pixel_4_5',
 'pixel_4_6',
 'pixel_4_7',
 'pixel_5_0',
 'pixel_5_1',
 'pixel_5_2',
 'pixel_5_3',
 'pixel_5_4',
 'pixel_5_5',
 'pixel_5_6',
 'pixel_5_7',
 'pixel_6_0',
 'pixel_6_1',
 'pixel_6_2',
 'pixel_6_3',
 'pixel_6_4',
 'pixel_6_5',
 'pixel_6_6',
 'pixel_6_7',
 'pixel_7_0',
 'pixel_7_1',
 'pixel_7_2',
 'pixel_7_3',
 'pixel_7_4',
 'pixel_7_5',
 'pixel_7_6',
 'pixel_7_7']

#### (7) 2차원 배열 데이터를 표로 만들어보자! (pandas 라이브러리 이용)

In [8]:
# 판다스 불러와서 버전 확인
import pandas as pd
print(pd.__version__)

# 손글씨 데이터셋을 표로 변환 = DataFrame이라는 자료형으로 변환
digits_df = pd.DataFrame(data=digits_data, columns=digits.feature_names)
digits_df

1.1.5


pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        5.0        0.0   
3           0.0        0.0        0.0        8.0  ...        9.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        4.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        1.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        2.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        8.0        0.0   

      pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0           0.0        0.0        6.0       13.0       10.0        0.0   
1           0.0        0.0        0.0       11.0       16.0       10.0   
2           0.0        0.0        0.0        3.0       11.0       16.0   
3           0.0        0.0        7.0       13.0       13.0        9.0   
4           0.0        0.0        0.0        2.0       16.0        4.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        2.0       14.0       15.0        9.0   
1793        0.0        0.0        6.0       16.0       14.0        6.0   
1794        0.0        0.0        2.0        9.0       13.0        6.0   
1795        0.0        0.0        5.0       12.0       16.0       12.0   
1796        0.0        1.0        8.0       12.0       14.0       12.0   

      pixel_7_6  pixel_7_7  
0           0.0        0.0  
1           0.0        0.0  
2           9.0        0.0  
3           0.0        0.0  
4           0.0        0.0  
...         ...        ...  
1792        0.0        0.0  
1793        0.0        0.0  
1794        0.0        0.0  
1795        0.0        0.0  
1796        1.0        0.0  

[1797 rows x 64 columns]

#### (8) 마지막 열에 정답 데이터 추가 ($\rightarrow$ 추가된 것은 표 아래의 행렬 사이즈로 확인!)

In [9]:
digits_df["label"] = digits.target
digits_df

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_7  pixel_7_0  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        0.0        0.0   
3           0.0        0.0        0.0        8.0  ...        0.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        0.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        0.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        0.0        0.0   

      pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  pixel_7_6  \
0           0.0        6.0       13.0       10.0        0.0        0.0   
1           0.0        0.0       11.0       16.0       10.0        0.0   
2           0.0        0.0        3.0       11.0       16.0        9.0   
3           0.0        7.0       13.0       13.0        9.0        0.0   
4           0.0        0.0        2.0       16.0        4.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        2.0       14.0       15.0        9.0        0.0   
1793        0.0        6.0       16.0       14.0        6.0        0.0   
1794        0.0        2.0        9.0       13.0        6.0        0.0   
1795        0.0        5.0       12.0       16.0       12.0        0.0   
1796        1.0        8.0       12.0       14.0       12.0        1.0   

      pixel_7_7  label  
0           0.0      0  
1           0.0      1  
2           0.0      2  
3           0.0      3  
4           0.0      4  
...         ...    ...  
1792        0.0      9  
1793        0.0      0  
1794        0.0      8  
1795        0.0      9  
1796        0.0      8  

[1797 rows x 65 columns]

### 3. 데이터 분리

: 테스트의 공정성을 위해 학습데이터셋과 테스트데이터셋을 분리시킴 (사이킷런이 제공하는 `train_test_split` 함수 이용)

#### (1) 데이터 분리, 테스트데이터 사이즈 조절 및 랜덤성 결정

In [10]:
X_train, X_test, y_train, y_test = train_test_split(digits_data, digits_label,
                                                    test_size = 0.2,  # 전체 1797 중의 0.2% = 1797*0.2 = 359.4
                                                    random_state = None)
print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

X_train 개수:  1437 , X_test 개수:  360


#### (2) 데이터셋 확인

In [11]:
X_train.shape, y_train.shape

((1437, 64), (1437,))

In [12]:
X_test.shape, y_test.shape

((360, 64), (360,))

In [13]:
#y-데이터 랜덤성 확인 → 0~9까지의 라벨이 잘 섞였는가?? Yes!!!

y_train, y_test

(array([9, 4, 6, ..., 1, 6, 7]),
 array([0, 6, 1, 8, 1, 5, 5, 6, 8, 0, 6, 6, 4, 5, 3, 4, 9, 2, 5, 2, 3, 9,
        3, 7, 8, 2, 0, 9, 5, 8, 9, 3, 1, 3, 4, 2, 7, 6, 6, 9, 6, 9, 3, 1,
        3, 9, 6, 4, 6, 7, 9, 5, 4, 4, 9, 4, 2, 5, 6, 0, 2, 7, 5, 9, 4, 7,
        9, 9, 7, 3, 3, 7, 2, 6, 4, 1, 2, 4, 3, 1, 6, 3, 2, 4, 8, 9, 7, 3,
        9, 1, 1, 1, 0, 6, 8, 5, 3, 9, 9, 1, 0, 3, 6, 0, 2, 2, 9, 4, 4, 4,
        2, 7, 8, 7, 8, 5, 6, 5, 7, 8, 2, 0, 3, 3, 3, 9, 5, 2, 1, 4, 8, 2,
        0, 1, 4, 2, 0, 1, 0, 1, 5, 2, 2, 9, 7, 2, 5, 5, 9, 4, 0, 5, 2, 8,
        7, 8, 2, 0, 8, 8, 9, 2, 2, 0, 0, 8, 4, 6, 6, 4, 2, 4, 7, 9, 9, 5,
        3, 8, 6, 8, 3, 7, 2, 5, 9, 7, 7, 3, 8, 4, 7, 9, 4, 7, 0, 0, 3, 5,
        3, 0, 6, 6, 5, 8, 4, 2, 1, 0, 6, 7, 5, 9, 9, 6, 5, 3, 2, 7, 6, 9,
        6, 2, 9, 0, 2, 9, 9, 3, 4, 4, 6, 2, 3, 5, 0, 4, 6, 5, 1, 0, 4, 1,
        0, 1, 3, 9, 3, 9, 3, 7, 8, 8, 6, 0, 4, 7, 5, 8, 6, 2, 4, 4, 9, 7,
        5, 0, 0, 7, 6, 4, 4, 6, 0, 4, 7, 1, 3, 4, 4, 5, 4, 8, 8, 7, 6, 9,
     

### 4. 모델 학습 및 예측

#### (1) Decision Tree 이용한 분류 및 예측

In [14]:
# 의사결정트리 변수에 모델 저장
decision_tree = DecisionTreeClassifier(random_state=None)
print(decision_tree._estimator_type)

# 모델 학습
decision_tree.fit(X_train, y_train)

# 예측
y_pred = decision_tree.predict(X_test)

# 정답과 예측값 출력
print(classification_report(y_test, y_pred))

classifier
              precision    recall  f1-score   support

           0       0.97      0.89      0.93        35
           1       0.82      0.96      0.88        24
           2       0.87      0.85      0.86        39
           3       0.80      0.76      0.78        37
           4       0.83      0.83      0.83        46
           5       0.84      0.91      0.88        35
           6       1.00      0.90      0.95        39
           7       0.76      0.81      0.78        31
           8       0.82      0.84      0.83        32
           9       0.76      0.76      0.76        42

    accuracy                           0.84       360
   macro avg       0.85      0.85      0.85       360
weighted avg       0.85      0.84      0.85       360



In [15]:
# 오차 행렬
confusion_matrix(y_test, y_pred)

array([[31,  0,  1,  0,  0,  2,  0,  0,  0,  1],
       [ 0, 23,  0,  0,  0,  0,  0,  1,  0,  0],
       [ 0,  0, 33,  3,  1,  0,  0,  1,  1,  0],
       [ 0,  0,  3, 28,  0,  1,  0,  2,  0,  3],
       [ 0,  3,  0,  0, 38,  1,  0,  1,  1,  2],
       [ 0,  0,  0,  0,  1, 32,  0,  0,  1,  1],
       [ 0,  1,  0,  0,  3,  0, 35,  0,  0,  0],
       [ 1,  0,  0,  0,  3,  1,  0, 25,  0,  1],
       [ 0,  1,  0,  0,  0,  0,  0,  2, 27,  2],
       [ 0,  0,  1,  4,  0,  1,  0,  1,  3, 32]])

#### (2) Random Forest 이용한 분류 및 예측

In [16]:
random_forest = RandomForestClassifier(random_state = None)  # 변수에 모델 저장

random_forest.fit(X_train, y_train)  # 학습
y_ped = random_forest.predict(X_test)  # 예측

print(classification_report(y_test, y_pred))  # 예측 결과 출력

              precision    recall  f1-score   support

           0       0.97      0.89      0.93        35
           1       0.82      0.96      0.88        24
           2       0.87      0.85      0.86        39
           3       0.80      0.76      0.78        37
           4       0.83      0.83      0.83        46
           5       0.84      0.91      0.88        35
           6       1.00      0.90      0.95        39
           7       0.76      0.81      0.78        31
           8       0.82      0.84      0.83        32
           9       0.76      0.76      0.76        42

    accuracy                           0.84       360
   macro avg       0.85      0.85      0.85       360
weighted avg       0.85      0.84      0.85       360



In [17]:
# 오차 행렬
confusion_matrix(y_test, y_pred)

array([[31,  0,  1,  0,  0,  2,  0,  0,  0,  1],
       [ 0, 23,  0,  0,  0,  0,  0,  1,  0,  0],
       [ 0,  0, 33,  3,  1,  0,  0,  1,  1,  0],
       [ 0,  0,  3, 28,  0,  1,  0,  2,  0,  3],
       [ 0,  3,  0,  0, 38,  1,  0,  1,  1,  2],
       [ 0,  0,  0,  0,  1, 32,  0,  0,  1,  1],
       [ 0,  1,  0,  0,  3,  0, 35,  0,  0,  0],
       [ 1,  0,  0,  0,  3,  1,  0, 25,  0,  1],
       [ 0,  1,  0,  0,  0,  0,  0,  2, 27,  2],
       [ 0,  0,  1,  4,  0,  1,  0,  1,  3, 32]])

#### (3) SVM 이용한 분류 및 예측

In [18]:
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.97      0.99        35
           1       1.00      1.00      1.00        24
           2       1.00      1.00      1.00        39
           3       1.00      1.00      1.00        37
           4       0.98      0.98      0.98        46
           5       1.00      0.97      0.99        35
           6       1.00      1.00      1.00        39
           7       1.00      0.97      0.98        31
           8       0.97      1.00      0.98        32
           9       0.93      0.98      0.95        42

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



In [19]:
# 오차 행렬
confusion_matrix(y_test, y_pred)

array([[34,  0,  0,  0,  1,  0,  0,  0,  0,  0],
       [ 0, 24,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 39,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 37,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0,  0, 45,  0,  0,  0,  0,  1],
       [ 0,  0,  0,  0,  0, 34,  0,  0,  0,  1],
       [ 0,  0,  0,  0,  0,  0, 39,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 30,  0,  1],
       [ 0,  0,  0,  0,  0,  0,  0,  0, 32,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  1, 41]])

#### (4) SGD Classifier 이용한 분류 및 예측

In [20]:
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        35
           1       0.77      1.00      0.87        24
           2       1.00      1.00      1.00        39
           3       0.90      1.00      0.95        37
           4       1.00      0.91      0.95        46
           5       1.00      0.86      0.92        35
           6       0.95      1.00      0.97        39
           7       1.00      0.94      0.97        31
           8       1.00      0.72      0.84        32
           9       0.84      0.98      0.90        42

    accuracy                           0.94       360
   macro avg       0.95      0.94      0.94       360
weighted avg       0.95      0.94      0.94       360



In [21]:
# 오차 행렬
confusion_matrix(y_test, y_pred)

array([[35,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 24,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 39,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 37,  0,  0,  0,  0,  0,  0],
       [ 0,  2,  0,  0, 42,  0,  1,  0,  0,  1],
       [ 0,  2,  0,  1,  0, 30,  0,  0,  0,  2],
       [ 0,  0,  0,  0,  0,  0, 39,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 29,  0,  2],
       [ 0,  3,  0,  2,  0,  0,  1,  0, 23,  3],
       [ 0,  0,  0,  1,  0,  0,  0,  0,  0, 41]])

#### (5) Logistic Regression 이용한 분류 및 예측

In [22]:
logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      1.00      0.99        35
           1       0.83      1.00      0.91        24
           2       1.00      1.00      1.00        39
           3       0.97      1.00      0.99        37
           4       1.00      0.96      0.98        46
           5       1.00      0.89      0.94        35
           6       1.00      0.97      0.99        39
           7       1.00      0.94      0.97        31
           8       0.97      0.97      0.97        32
           9       0.93      0.98      0.95        42

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [23]:
# 오차 행렬
confusion_matrix(y_test, y_pred)

array([[35,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0, 24,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0, 39,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  0,  0, 37,  0,  0,  0,  0,  0,  0],
       [ 0,  2,  0,  0, 44,  0,  0,  0,  0,  0],
       [ 1,  1,  0,  1,  0, 31,  0,  0,  0,  1],
       [ 0,  1,  0,  0,  0,  0, 38,  0,  0,  0],
       [ 0,  0,  0,  0,  0,  0,  0, 29,  0,  2],
       [ 0,  1,  0,  0,  0,  0,  0,  0, 31,  0],
       [ 0,  0,  0,  0,  0,  0,  0,  0,  1, 41]])

### $4^{\prime}$. 다양한 평가 척도의 필요성과 오차행렬 (이 부분은 개념 이해를 위한 설명)

#### (1) 불균형 데이터에 대한 문제 파악

* 숫자 10개를 분류하는 것이 아닌, 해당 이미지가 3인지 아닌지 맞추는 문제로 변형하여 풀기  
* 즉, 입력 데이터가 3 이면 3 출력, 3 아닌면 0 출력
* 즉, 길이는 예측과 같고 모두 0으로 채워진 리스트를 이용

In [24]:
digits = load_digits()
digits.keys()

digits_data = digits.data
digits_data.shape

digits_label = digits.target

# 기존의 digits_label 에서 숫자가 3이면 그대로 3, 아니면 0 저장하는 새로운 라벨 만들기
new_label = [3 if i == 3 else 0 for i in digits_label]
new_label[:20]

# 새로운 라벨에 대하여 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data,
                                                    new_label,
                                                    test_size=0.2,
                                                    random_state=None)

# 새로운 라벨에 대하여 모델 학습
decision_tree = DecisionTreeClassifier(random_state=None)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

# 새로운 라벨에 대한 예측 및 정확도
accuracy = accuracy_score(y_test, y_pred)
accuracy

0.9611111111111111

(해석 : 3 이외의 다른 라벨들을 모두 0으로 바꾸었기 때문에, 정답 라벨이 0인 데이터가 96% 정도로 매우 높음)

#### (2) 0으로만 찍었을 때의 정확도 확인

In [25]:
fake_pred = [0] * len(y_pred)

accuracy = accuracy_score(y_test, fake_pred)
accuracy

0.8916666666666667

(해석 : 0으로만 찍어도 무려 90% 정도의 정확도. 이로써 (1)의 결과 확인! 불균형 데이터에서 자주 발생하는 문제)

###### "정확도(Accuracy)는 정답의 분포에 따라 모델의 성능을 제대로 평가하지 못하는 척도일 수 있다."

#### (3) 해결 : 정답 뿐 아니라 오답도 확인할 수 있는 "오차행렬" 이용!!

In [26]:
# 모델이 예측한 손글씨 결과
confusion_matrix(y_test, y_pred)

array([[315,   6],
       [  8,  31]])

오차행렬 Confusion Matrix 에서 대각성분은 예측값과 실제값이 일치하는 경우이고 모든 대각성분의 점수가 높게 나왔다.  
즉, 이 분류 모델은 예측이 거의 정확하다는 의미!!

In [27]:
# 모든 숫자(데이터)를 0(=positive)으로 예측한 fake 예측 결과 : 그래서 마지막열인 FN, TN 모두 0 !!
confusion_matrix(y_test, fake_pred)

array([[321,   0],
       [ 39,   0]])

In [28]:
# 정밀도 Precision, 재현율 Recall, f1-score, 정확도 Accuracy 한눈에 보기
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.98      0.98      0.98       321
           3       0.84      0.79      0.82        39

    accuracy                           0.96       360
   macro avg       0.91      0.89      0.90       360
weighted avg       0.96      0.96      0.96       360



(해석 : 0은 324개로 많기 때문에 정밀도/재현율 모두 높은 점수, 3의 경우 36개로 적은 수이기 때문인지 정밀도와 재현율이 다소 낮음)

In [29]:
# fake 인 경우도 한눈에 확인
print(classification_report(y_test, fake_pred))

              precision    recall  f1-score   support

           0       0.89      1.00      0.94       321
           3       0.00      0.00      0.00        39

    accuracy                           0.89       360
   macro avg       0.45      0.50      0.47       360
weighted avg       0.80      0.89      0.84       360



/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


(해석 : 0에 대한 정밀도/재현율은 굉장히 높지만 3에 대한 정밀도/재현율은 0. 즉, 0은 아주 잘 예측하지만 3은 하나도 맞추지 못함을 의미)

In [30]:
# 정확도 확인
accuracy_score(y_test, y_pred), accuracy_score(y_test, fake_pred)

(0.9611111111111111, 0.8916666666666667)

(해석 : 위에서 이미 확인한 바이지만, 정확도는 이처럼 큰 차이가 없어서 라벨이 불균형하게 분포되어 있는 데이터의 경우 오판 가능! 이럴 때 정밀도와 재현율이 필요!!)

### 5. 모델별 비교 평가

* acc.=accuracy
* m-avg(p,r)=macro avg(precision, recall)
* w-avg(p,r)=weighted avg(precision, recall)
  
|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|acc. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|m-avg(p,r) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|w-avg(p,r) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|
|:---------------|:----------|:----------|:----------|
|Decision Tree|0.85|(0.85, 0.84)|(0.86, 0.85)|
|Random Forest|0.85|(0.85, 0.84)|(0.86, 0.85)|
|SVM-model|0.99|(0.99, 0.98)|(0.99, 0.99)|
|SGD-model|0.95|(0.95, 0.95)|(0.96, 0.95)|
|Logistic Regression|0.97|(0.97, 0.97)|(0.97, 0.97)|

위의 오차행렬 및 각 척도의 분석 결과, 각 모델마다 모든 라벨의 "정밀도 precision, 재현율 recall"은 골고루 큰 수치가 나왔으므로 데이터는 비교적 균형있게 분포되어 있다고 판단된다.  
따라서, "정확도 Accuracy"로써 각 모델의 성능을 비교하는 것이 유의미하다고 보여진다.
  
요컨대, 손글씨 분류 모델 중 가장 높은 정확도를 보인 모델은 SVM-model 이며(SGD 모델과 로지스틱회귀 모델 역시 비슷한 수치로 높음) 의사결정나무와 랜덤포레스트의 정확도는 비교적 낮은 수치를 보인다.
  
한편, random_state 값에 변화를 주었을 때 SGD-model과 로지스틱 모델의 정확도는 다소 상승하기도 하며 다른 모델들의 수치 변화는 미미했는데 이것의 의미를 아직 이해하지 못했다.
이러한 이유로 노드에 명시된 랜덤 스테이트값을 이용하지 않고, np.random에서 제공하는 random number generator가 사용되도록 None을 입력 (자세한 의미는 모름)

## <span style="color:blue">프로젝트-2 load_wine : 와인 분류</span>

### 1. 필요한 모듈 import하기

In [31]:
# 데이터 준비, 데이터 분리 관련 모듈들
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split

# 데이터 학습 모델들(5가지) : 의사결정트리, 랜덤포레스트, SVM, SGD, 로지스틱회귀
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

# 모델 평가를 위한 오차행렬과 각 지표를 한눈에 확인하기 위한 모듈
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score  # 노드에는 이에 대한 언급 없이 그냥 사용해서, 코드를 그대로 따라하면 오류 발생

### 2. 데이터 준비

#### (1) 데이터셋 가져오기

In [32]:
wine = load_wine()

print(type(dir(wine)))  # dir() = 객체의 변수와 메서드 나열
wine.keys()  # keys() 메서드 = wine 에 담긴 정보 확인

<class 'list'>


dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

#### (2) 데이터를 손글씨 변수에 저장하고 크기 확인하기

In [33]:
# shape는 배열의 형상정보를 출력
#이 경우, 178개의 데이터가 13개의 정보를 담고 있음. 아래의 (5)DESCR 참고

wine_data = wine.data
print(wine_data.shape)

(178, 13)


#### (3) 라벨 변수에 타겟 저장

In [34]:
wine_label = wine.target  # 변수에 타겟 저장
print(wine_label.shape)  # 타겟의 길이와 형태 확인 : 총 178개의 데이터, 라벨 3가지
wine_label

(178,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

#### (4) 타겟 이름 출력

In [35]:
wine.target_names  # 3가지로 분류되니까 0~2까지 순서대로 라벨 출력

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

#### (5) 데이터셋 설명 확인

In [36]:
print(wine.DESCR)  # 말그대로 describe

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

#### (6) feature 이름 확인

In [37]:
wine.feature_names  # (2)shape 와 (5) 참고, feature 정보 13가지였음

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

#### (7) 2차원 배열 데이터를 표로 만들어보자! (pandas 라이브러리 이용)

In [38]:
# 판다스 불러와서 버전 확인
import pandas as pd
print(pd.__version__)

# 와인 데이터셋을 표로 변환 = DataFrame이라는 자료형으로 변환
wine_df = pd.DataFrame(data=wine_data, columns=wine.feature_names)
wine_df

1.1.5


alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

#### (8) 마지막 열에 정답 데이터 추가 ($\rightarrow$ 추가된 것은 표 아래의 행렬 사이즈로 확인!)

In [39]:
wine_df["label"] = wine.target
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  label  
0                            3.92   1065.0      0  
1                            3.40   1050.0      0  
2                            3.17   1185.0      0  
3                            3.45   1480.0      0  
4                            2.93    735.0      0  
..                            ...      ...    ...  
173                          1.74    740.0      2  
174                          1.56    750.0      2  
175                          1.56    835.0      2  
176                          1.62    840.0      2  
177                          1.60    560.0      2  

[178 rows x 14 columns]

### 3. 데이터 분리

: 테스트의 공정성을 위해 학습데이터셋과 테스트데이터셋을 분리시킴 (사이킷런이 제공하는 `train_test_split` 함수 이용)

#### (1) 데이터 분리, 테스트데이터 사이즈 조절 및 랜덤성 결정

In [40]:
X_train, X_test, y_train, y_test = train_test_split(wine_data, wine_label,
                                                    test_size = 0.2,  # 전체 178 중의 0.2% = 178*0.2 = 35.6
                                                    random_state = None)
print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

X_train 개수:  142 , X_test 개수:  36


#### (2) 데이터셋 확인

In [41]:
X_train.shape, y_train.shape

((142, 13), (142,))

In [42]:
X_test.shape, y_test.shape

((36, 13), (36,))

In [43]:
#y-데이터 랜덤성 확인 → 0~2까지의 라벨이 잘 섞였는가?? Yes!!!

y_train, y_test

(array([0, 1, 1, 0, 0, 2, 2, 0, 1, 2, 1, 2, 0, 1, 2, 1, 0, 1, 0, 1, 0, 1,
        0, 2, 2, 2, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 2, 1, 2, 1, 0, 1,
        2, 1, 0, 1, 1, 2, 1, 0, 0, 0, 0, 1, 2, 0, 0, 1, 0, 1, 0, 1, 2, 1,
        1, 2, 2, 2, 1, 1, 1, 0, 1, 0, 2, 2, 2, 1, 0, 2, 1, 0, 0, 2, 0, 1,
        0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 2, 0, 2, 1, 0, 2, 0, 2, 1, 1,
        0, 2, 1, 1, 1, 2, 1, 1, 2, 0, 2, 1, 2, 1, 1, 0, 0, 2, 2, 1, 1, 1,
        2, 1, 2, 1, 1, 0, 0, 2, 2, 1]),
 array([0, 0, 0, 0, 0, 1, 2, 1, 0, 0, 1, 1, 1, 1, 2, 0, 2, 1, 2, 0, 2, 1,
        0, 2, 1, 2, 1, 0, 1, 2, 2, 2, 0, 2, 0, 0]))

### 4. 모델 학습 및 예측

#### (1) Decision Tree 이용한 분류 및 예측

In [44]:
# 의사결정트리 변수에 모델 저장
decision_tree = DecisionTreeClassifier(random_state=None)
print(decision_tree._estimator_type)

# 모델 학습
decision_tree.fit(X_train, y_train)

# 예측
y_pred = decision_tree.predict(X_test)

# 정답과 예측값 출력
print(classification_report(y_test, y_pred))

classifier
              precision    recall  f1-score   support

           0       1.00      0.93      0.96        14
           1       0.77      0.91      0.83        11
           2       0.90      0.82      0.86        11

    accuracy                           0.89        36
   macro avg       0.89      0.89      0.88        36
weighted avg       0.90      0.89      0.89        36



In [45]:
confusion_matrix(y_test, y_pred)

array([[13,  1,  0],
       [ 0, 10,  1],
       [ 0,  2,  9]])

#### (2) Random Forest 이용한 분류 및 예측

In [46]:
random_forest = RandomForestClassifier(random_state = None)  # 변수에 모델 저장

random_forest.fit(X_train, y_train)  # 학습
y_ped = random_forest.predict(X_test)  # 예측

print(classification_report(y_test, y_pred))  # 예측 결과 출력

              precision    recall  f1-score   support

           0       1.00      0.93      0.96        14
           1       0.77      0.91      0.83        11
           2       0.90      0.82      0.86        11

    accuracy                           0.89        36
   macro avg       0.89      0.89      0.88        36
weighted avg       0.90      0.89      0.89        36



In [47]:
confusion_matrix(y_test, y_pred)

array([[13,  1,  0],
       [ 0, 10,  1],
       [ 0,  2,  9]])

#### (3) SVM 이용한 분류 및 예측

In [48]:
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.85      0.79      0.81        14
           1       0.48      1.00      0.65        11
           2       0.00      0.00      0.00        11

    accuracy                           0.61        36
   macro avg       0.44      0.60      0.49        36
weighted avg       0.48      0.61      0.51        36



/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [49]:
confusion_matrix(y_test, y_pred)

array([[11,  3,  0],
       [ 0, 11,  0],
       [ 2,  9,  0]])

#### (4) SGD Classifier 이용한 분류 및 예측

In [50]:
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.14      0.25        14
           1       0.00      0.00      0.00        11
           2       0.32      1.00      0.49        11

    accuracy                           0.36        36
   macro avg       0.44      0.38      0.25        36
weighted avg       0.49      0.36      0.25        36



/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [51]:
# 오차 행렬
confusion_matrix(y_test, y_pred)

array([[ 2,  0, 12],
       [ 0,  0, 11],
       [ 0,  0, 11]])

#### (5) Logistic Regression 이용한 분류 및 예측

In [52]:
logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       0.92      1.00      0.96        11
           2       1.00      0.91      0.95        11

    accuracy                           0.97        36
   macro avg       0.97      0.97      0.97        36
weighted avg       0.97      0.97      0.97        36



/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [53]:
# 오차 행렬
confusion_matrix(y_test, y_pred)

array([[14,  0,  0],
       [ 0, 11,  0],
       [ 0,  1, 10]])

### 5. 모델별 비교 평가

* acc.=accuracy
* m-avg(p,r)=macro avg(precision, recall)
* w-avg(p,r)=weighted avg(precision, recall)
  
|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|acc. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|m-avg(p,r) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|w-avg(p,r) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|
|:---------------|:----------|:----------|:----------|
|Decision Tree|0.97|(0.98, 0.98)|(0.97, 0.97)|
|Random Forest|0.97|(0.98, 0.98)|(0.97, 0.97)|
|SVM-model|0.75|(0.64, 0.66)|(0.69, 0.75)|
|SGD-model|0.50|(0.77, 0.57)|(0.85, 0.50)|
|Logistic Regression|0.97|(0.98, 0.98)|(0.97, 0.97)|

(1)의사결정트리, (2)랜덤포레스트, (5)로지스틱회귀 모델들은 모든 라벨의 "정밀도 precision, 재현율 recall"이 골고루 큰 수치가 나왔으며, 정확도 역시 세 모델 모두 0.97로 높게 나왔다.  
세 모델의 각 평가 척도의 차이는 없으므로 와인분류모델은 선호도에 따라 결정해도 무방하다고 생각한다.
  
한편, SVM 과 SGD 두 모델 모두 "라벨 2"에 대한 정밀도가 현저하게 낮으며 정확도 역시 낮은 수치로 나타났기 때문에 이 두 모델은 와인분류문제에 적합하지 않은 것으로 판단된다.

## <span style="color:blue">프로젝트-3 load_breast_cancer : 유방암 여부 진단</span>

### 1. 필요한 모듈 import하기

In [54]:
# 데이터 준비, 데이터 분리 관련 모듈들
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

# 데이터 학습 모델들(5가지) : 의사결정트리, 랜덤포레스트, SVM, SGD, 로지스틱회귀
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

# 모델 평가를 위한 오차행렬과 각 지표를 한눈에 확인하기 위한 모듈
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score  # 노드에는 이에 대한 언급 없이 그냥 사용해서, 코드를 그대로 따라하면 오류 발생

### 2. 데이터 준비

#### (1) 데이터셋 가져오기

In [55]:
breast_cancer = load_breast_cancer()

print(type(dir(breast_cancer)))  # dir() = 객체의 변수와 메서드 나열
breast_cancer.keys()  # keys() 메서드 = breast_cancer 에 담긴 정보 확인

<class 'list'>


dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])

#### (2) 데이터를 손글씨 변수에 저장하고 크기 확인하기

In [56]:
# shape는 배열의 형상정보를 출력
#이 경우, 569개의 데이터가 30개의 정보를 담고 있음. 아래의 (5)DESCR 참고

breast_cancer_data = breast_cancer.data
print(breast_cancer_data.shape)

(569, 30)


#### (3) 라벨 변수에 타겟 저장

In [57]:
breast_cancer_label = breast_cancer.target  # 변수에 타겟 저장
print(breast_cancer_label.shape)  # 타겟의 길이와 형태 확인 : 총 569개의 데이터, 라벨 2가지
breast_cancer_label

(569,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

#### (4) 타겟 이름 출력

In [58]:
breast_cancer.target_names  # 2가지로 분류되니까 0, 1 순서대로 라벨 출력

array(['malignant', 'benign'], dtype='<U9')

#### (5) 데이터셋 설명 확인

In [59]:
print(breast_cancer.DESCR)  # 말그대로 describe

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

#### (6) feature 이름 확인

In [60]:
breast_cancer.feature_names  # (2)shape 와 (5) 참고, feature 정보 30가지였음

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

#### (7) 2차원 배열 데이터를 표로 만들어보자! (pandas 라이브러리 이용)

In [61]:
# 판다스 불러와서 버전 확인
import pandas as pd
print(pd.__version__)

# 유방암 데이터셋을 표로 변환 = DataFrame이라는 자료형으로 변환
breast_cancer_df = pd.DataFrame(data=breast_cancer_data, columns=breast_cancer.feature_names)
breast_cancer_df

1.1.5


mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

#### (8) 마지막 열에 정답 데이터 추가 ($\rightarrow$ 추가된 것은 표 아래의 행렬 사이즈로 확인!)

In [62]:
breast_cancer_df["label"] = breast_cancer.target
breast_cancer_df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  label  
0                  0.2654          0.4601                  0.11890      0  
1                  0.1860          0.2750                  0.08902      0  
2                  0.2430          0.3613                  0.08758      0  
3                  0.2575          0.6638                  0.17300      0  
4                  0.1625          0.2364                  0.07678   

### 3. 데이터 분리

: 테스트의 공정성을 위해 학습데이터셋과 테스트데이터셋을 분리시킴 (사이킷런이 제공하는 `train_test_split` 함수 이용)

#### (1) 데이터 분리, 테스트데이터 사이즈 조절 및 랜덤성 결정

In [63]:
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data, breast_cancer_label,
                                                    test_size = 0.2,  # 전체 569 중의 0.2% = 569*0.2 = 113.8
                                                    random_state = None)
print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

X_train 개수:  455 , X_test 개수:  114


#### (2) 데이터셋 확인

In [64]:
X_train.shape, y_train.shape

((455, 30), (455,))

In [65]:
X_test.shape, y_test.shape

((114, 30), (114,))

In [66]:
#y-데이터 랜덤성 확인 → 0~2까지의 라벨이 잘 섞였는가?? Yes!!!

y_train, y_test

(array([1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0,
        1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1,
        1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1,
        1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1,
        1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0,
        1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1,
        0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0,
        1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1,
        1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0,
        1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1,
        1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
        0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 

### 4. 모델 학습 및 예측

#### (1) Decision Tree 이용한 분류 및 예측

In [67]:
# 의사결정트리 변수에 모델 저장
decision_tree = DecisionTreeClassifier(random_state=None)
print(decision_tree._estimator_type)

# 모델 학습
decision_tree.fit(X_train, y_train)

# 예측
y_pred = decision_tree.predict(X_test)

# 정답과 예측값 출력
print(classification_report(y_test, y_pred))

classifier
              precision    recall  f1-score   support

           0       0.89      0.82      0.85        39
           1       0.91      0.95      0.93        75

    accuracy                           0.90       114
   macro avg       0.90      0.88      0.89       114
weighted avg       0.90      0.90      0.90       114



In [68]:
confusion_matrix(y_test, y_pred)

array([[32,  7],
       [ 4, 71]])

#### (2) Random Forest 이용한 분류 및 예측

In [69]:
random_forest = RandomForestClassifier(random_state = None)  # 변수에 모델 저장

random_forest.fit(X_train, y_train)  # 학습
y_ped = random_forest.predict(X_test)  # 예측

print(classification_report(y_test, y_pred))  # 예측 결과 출력

              precision    recall  f1-score   support

           0       0.89      0.82      0.85        39
           1       0.91      0.95      0.93        75

    accuracy                           0.90       114
   macro avg       0.90      0.88      0.89       114
weighted avg       0.90      0.90      0.90       114



In [70]:
confusion_matrix(y_test, y_pred)

array([[32,  7],
       [ 4, 71]])

#### (3) SVM 이용한 분류 및 예측

In [71]:
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.93      0.69      0.79        39
           1       0.86      0.97      0.91        75

    accuracy                           0.88       114
   macro avg       0.89      0.83      0.85       114
weighted avg       0.88      0.88      0.87       114



In [72]:
confusion_matrix(y_test, y_pred)

array([[27, 12],
       [ 2, 73]])

#### (4) SGD Classifier 이용한 분류 및 예측

In [73]:
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.72      0.84        39
           1       0.87      1.00      0.93        75

    accuracy                           0.90       114
   macro avg       0.94      0.86      0.88       114
weighted avg       0.92      0.90      0.90       114



In [74]:
confusion_matrix(y_test, y_pred)

array([[28, 11],
       [ 0, 75]])

#### (5) Logistic Regression 이용한 분류 및 예측

In [75]:
logistic_model = LogisticRegression()

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.74      0.81        39
           1       0.88      0.95      0.91        75

    accuracy                           0.88       114
   macro avg       0.88      0.85      0.86       114
weighted avg       0.88      0.88      0.87       114



/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [76]:
confusion_matrix(y_test, y_pred)

array([[29, 10],
       [ 4, 71]])

### 5. 모델별 비교 평가

* acc.=accuracy
* m-avg(p,r)=macro avg(precision, recall)
* w-avg(p,r)=weighted avg(precision, recall)
  
|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|acc. &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|m-avg(p,r) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|w-avg(p,r) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|
|:---------------|:----------|:----------|:----------|
|Decision Tree|0.94|(0.93, 0.93)|(0.94, 0.94)|
|Random Forest|0.94|(0.93, 0.93)|(0.94, 0.94)|
|SVM-model|0.89|(0.91, 0.86)|(0.90, 0.89)|
|SGD-model|0.87|(0.92, 0.80)|(0.89, 0.87)|
|Logistic Regression|0.92|(0.92, 0.90)|(0.92, 0.92)|

(1)의사결정트리, (2)랜덤포레스트, (5)로지스틱회귀 모델들은 모든 라벨의 "정밀도 precision, 재현율 recall"이 골고루 큰 수치가 나왔으며, 정확도 역시 세 모델 모두 0.92 이상으로 높게 나왔다.  
이 결과에서 세 모델의 각 평가 척도의 차이는 거의 없으나, 유방암여부진단 문제만큼은 randon_state 값에 따라 모델 성능 결과가 사뭇 다르게 도출되는 것으로 보인다.  
특히, '랜덤=32'로 설정했을 때에는 "로지스틱회귀" 모델의 정확도가 제일 높게 나오기도 한다.
인간의 생명과 연관된 문제인만큼 좀 더 정확한 예측을 할 수 있는 연구와 개발이 활발히 이루어지기를 바란다.